README.md
# CommonJS `require` in 1k, with no build needed

This project implements a minimal, and yet practically useful, CommonJS/Node.js `require` module loader for the browser in under 1000 bytes.

## Features

* Synchronous `require` in modules
* `exports` and `module.exports` object
* Loaded modules are debuggable as normal in Chrome, Opera 15+, Firefox 36+ and Safari with `debugger` statement. (IE to be tested.)
* Works with npm, by searching `node_modules` directories (see Limitations)
* Handles circular dependencies
* Just runs, no need to build your JavaScript
* A pretty rich API, considering:
* it's 1k!

## Examples

* [Simple](http://stuk.github.io/require1k/examples/simple/)
* [Marked](http://stuk.github.io/require1k/examples/marked) markdown parser
* [React](http://stuk.github.io/require1k/examples/react) framework (!!)

## Usage

In your modules you can use the `require` function, `exports` object and `module.exports` object as you would in Node. There are two ways to kick off the require system:

### script `data-main` attribute

```html
<script src="require1k.min.js" data-main="./index"></script>
```

On start require1k will search for the first `<script>` tag with a `data-main` attribute. If found the named module will be resolved against the location of the html file, asynchronously loaded and then executed.

### global `R(func, callback)`

```html
<script>
R(function (require, module, exports) {
    var index = require("./index");
    exports.hello = "World!"
}, function (err, exports) {
    if (err) {
        console.error(err.statusText);
        return;
    }

    console.log(exports.hello)
});
</script>
```

Require1k also adds a global function, `R`, that accepts either a function or a module ID and an optional `callback`. When given a function it will load all the dependencies of the function and then execute it. If there was an error the callback gets passed the XMLHttpRequest object that failed as the first argument. It also gets passed the `exports` of the function, which you may find useful.

### global `R(moduleId, callback)`

```html
<script>
R("./index", function (err, index) {
    if (err) {
        console.error(err.statusText);
        return;
    }
    // use the exports object as needed
});
</script>
```

Alternatively you can provide a module ID and a optional `callback`. The named module is resolved against the location of the HTML file. If there was an error the callback gets passed the XMLHttpRequest object that failed as the first argument, otherwise the `exports` of the module are passed as the second argument. The callback may be called synchronously if the module is already in the internal module cache.

## Using script[data-base]

You can use the attribute `data-base` to set the base directory to load the main file. It usefull when you want to load file from the root server.

```html
<script src="require1k.min.js" data-main="/scripts/lib" data-base="/"></script>
```

By default, require1k converts `/scripts/lib` to `http://example.com/node_modules/scripts/lib.js`. With `data-base="/"`, require1k transform the URL to `http://example.com/scripts/lib.js`. This feature is needed when you use the HTML5 History Mode.

## Limitations

* `package.json` files are not loaded, and so the `"main"` property is ignored. Modules inside packages must be requested by their full path, e.g. `var _ = require("underscore/underscore");`
* Cross-package dependencies are found by walking up the URL path until a successful request is returned. This means that
    * loading cross-package dependencies is slow because requests are made sequentially until the module is found
    * 404 errors appear in the console
* Missing cross-package dependencies will exceed the stack size and halt script execution.
* Callback functions are called synchronously if the data is already available. This releases [Zalgo](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony), but hey, it's 1k.
* Does not load `.json` files


## The point

The short answer: because it's a fun hack.

The the long answer?

Using `<script>` tags to manage the dependencies of your "complex" web app makes for an unmaintainable mess.

Yet that's what most people are doing today, and even what a large number of articles currently being published *still* recommend. People are writing large and sophisticated apps and then wasting their time managing dependencies by hand! There's just no need for it.

Package management ensures you have a system for adding and updating libraries to your project. It means that you can easily check if your libraries are out of date, and see exactly what libraries your project depends on. It means that you don't have to go spelunking around the web to find the site where you can report an issue or download an update for this random file you have in your `js/` directory. But module management is even more important.

Module management is where you actually say "load this code for my app". And when you do this with a (multitude of) `<script>` tags on your page you can no longer isolate your code, and say "this is what this bit of code depends on". Every library you load is global. And what happens when you write some JavaScript that needs an extra library? You have to go back to this completely different file, and add something to the `HTML`. This is nonsense.

What each part of your app depends on should be explicit. It should not require editing an unrelated file just to depend on a third party module.

If you're writing a web app of any significant size today, you owe it to yourself to manage your dependencies correctly.

This is what I recommend:

* Use npm to manage your client side dependencies
* Use [Browserify](http://browserify.org/) or [Webpack](http://webpack.github.io/) to bundle up your application

That's it.

(And yep, I've noticed the irony of saying this while telling people to embed require1k on a page. But this is just a joke. Kind of.)

## Thanks

This project was largely inspired by working on [Mr](https://github.com/montagejs/mr), a CommonJS module loader and companion bundler that definitely isn't 1k.

Thanks to [Tom Robinson](https://twitter.com/tlrobinson) for writing the original version, and [Kris Kowal](https://twitter.com/kriskowal) for teaching me so much.

## License

BSD


CHANGELOG.md
# Changelog

## v2.0.0

* Ignore `require` calls in comments (#16)
* Ignore `$require` and `object.require` calls (#16)

## v1.0.1

* Initial release


spec/index.js
console.log("loaded index.js");

var second = require("./second");
console.log(second.thing);

R("./three", function (err, exports) {
    console.log("R callback", err, exports);
});

require("dep/index");


test/index.js
console.log("loaded index.js");

var second = require("./second");
assert(second.thing === "asd", "required export should be correct");

R("./three", function (err, exports) {
    console.log("R callback");
    assert(err === undefined, "err in callback should be undefined");
    assert(exports === "three", "exports in callback should be correct");
});

require("dep/index");

require("./multi-require/index");

require("./comment/index");


spec/second.js
console.log("loaded second");

exports.thing = "asd";

require("./index");


require1k.js
/*jshint node:false, -W082, -W017 */
R = (function (document, undefined) {
    // Each module has the following properties (shorted to one letter to aid compression)
    // - g: booleany, loadinG, truthy if this module has been requested for loading
    //      before. Used to prevent the same module being loaded twice
    // - l: string, Location, the url location of this module
    // - t: string, Text, the text content of the module
    // - e: booleany, Error, truthy if there was an error (probably a 404) loading the module
    // - n: module object, Next, instead of using this module, use the object
    //      pointed to by this property. Used for dependencies in other packages
    // - f: function, Factory, a function to use instead of eval'ing module.t
    // - exports, object, the exports of the module!
    var MODULES = {};

    // By using a named "eval" most browsers will execute in the global scope.
    // http://www.davidflanagan.com/2010/12/global-eval-in.html
    var globalEval = eval;

    // this variable is reused for a number of things to reduce the repetition
    // of strings. In the end is becomes "exports"
    var tmp = "createElement",
        baseElement = document[tmp]("base"),
        relativeElement = document[tmp]("a");
    document.head.appendChild(baseElement);


    // Loads the given module and all of it dependencies, recursively
    // - module         The module object
    // - callback       Called when everything has been loaded
    // - parentLocation Location of the parent directory to look in. Only given
    // for non-relative dependencies
    // - id             The name of the dependency. Only used for non-relative
    // dependencies
    function deepLoad(module, callback, parentLocation, id) {
        // If this module is already loading then don't proceed.
        // This is a bug.
        // If a module is requested but not loaded then the module isn't ready,
        // but we callback as if it is. Oh well, 1k!
        if (module.g) {
            return callback(module.e, module);
        }

        var location = module.g = module.l;

        var request = new XMLHttpRequest();
        request.onload = function (deps, count) {
            if (request.status == 200 || module.t) {
                // Should really use an object and then Object.keys to avoid
                // duplicate dependencies. But that costs bytes.
                deps = [];
                (module.t = module.t || request.response).replace(/(\/\*[\w\W]*?\*\/|\/\/[^\n]*|[.$]r)|\brequire\s*\(\s*["']([^"']*)["']\s*\)/g, function (_, ignore, id) {
                    if (!ignore) deps.push(id);
                });
                count = deps.length;
                function loaded() {
                    // We call loaded straight away below in case there
                    // are no dependencies. Putting this check first
                    // and the decrement after saves us an `if` for that
                    // special case
                    if (!count--) {
                        callback(undefined, module);
                    }
                }
                deps.map(function (dep) {
                    deepLoad(
                        resolveModuleOrGetExports(module.l, dep),
                        loaded,
                        // If it doesn't begin with a ".", then we're searching
                        // node_modules, so pass in the info to make this
                        // possible
                        dep[0] != "." ? location + "/../" : undefined,
                        dep
                    );
                });
                loaded();
            } else {
                // parentLocation is only given if we're searching in node_modules
                if (parentLocation) {
                    // Recurse up the tree trying to find the dependency
                    // (generating 404s on the way)
                    deepLoad(
                        module.n = resolveModuleOrGetExports(parentLocation += "../", id),
                        callback,
                        parentLocation,
                        id
                    );
                } else {
                    module.e = request;
                    callback(request, module);
                }
            }
        };

        // If the module already has text because we're using a factory
        // function, then there's no need to load the file!
        if (module.t) {
            request.onload();
        } else {
            request.open("GET", location, true);
            request.send();
        }
    }

    // Save bytes by combining two functions
    // - resolveModule which resolves a given relative path against the given
    //   base, and returns an existing or new module object
    // - getExports which returns the existing exports or runs the factory to
    //   create the exports for a module
    function resolveModuleOrGetExports(baseOrModule, relative, resolved) {
        // This should really be after the relative check, but because we are
        // `throw`ing, it messes up the optimizations. If we are being called
        // as resolveModule then the string `base` won't have the `e` property,
        // so we're fine.
        if (baseOrModule.e) {
            throw baseOrModule.e;
        }

        // If 2 arguments are given, then we are resolving modules...
        if (relative) {
            baseElement.href = baseOrModule;
            // If the relative url doesn't begin with a ".", then it's
            // in node_modules
            relativeElement.href = relative[0] != "." ? "./node_modules/" + relative : relative;
            resolved = relativeElement.href + ".js";
            baseElement.href = "";
            return (MODULES[resolved] = MODULES[resolved] || {l: resolved});
        }

        // ...otherwise we are getting the exports

        // Is this module a redirect to another one?
        if (baseOrModule.n) {
            return resolveModuleOrGetExports(baseOrModule.n);
        }

        if (!baseOrModule[tmp]) {
            (baseOrModule.f || globalEval("(function(require,"+tmp+",module){" + baseOrModule.t + "\n})//# sourceURL=" + baseOrModule.l))(
                function require (id) {
                    return resolveModuleOrGetExports(resolveModuleOrGetExports(baseOrModule.l, id));
                }, // require
                baseOrModule[tmp] = {}, // exports
                baseOrModule // module
            );
        }

        return baseOrModule[tmp];
    }

    function R(id, callback) {
        // If id has a `call` property it is a function, so make a module with
        // a factory
        deepLoad(id.call ? {l: "", t: "" + id, f: id} : resolveModuleOrGetExports("", id), function (err, module) {
            try {
                id = resolveModuleOrGetExports(module);
            } catch (_err) {
                err = _err
            }
            if (callback) {
                callback(err, id);
            }
        });
    }

    tmp = document.querySelector("script[data-main]");
    if (tmp) {
        R(tmp.dataset.main);
    }
    tmp = "exports";

    return R;

}(document));


robin.js
/*jshint node:false, -W082, -W017 */
R = (function (global, document, undefined) {
    var MODULES = {};

    // By using a named "eval" most browsers will execute in the global scope.
    // http://www.davidflanagan.com/2010/12/global-eval-in.html
    var globalEval = eval;

    var tmp = "createElement",
        baseElement = document[tmp]("base"),
        relativeElement = document[tmp]("a");
    document.head.appendChild(baseElement);

    function resolve(base, relative, resolved) {
        baseElement.href = base;
        // If the relative url begins with a letter (and not a "."), then it's
        // in node_modules
        relativeElement.href = relative.replace(/^(\w)/, "./node_modules/$1");
        resolved = relativeElement.href;
        baseElement.href = "";
        return resolved;

    }

    // A module has the following properties (shorted to one letter to aid compression)
    // - g: booleany, loadinG, truthy if this module has been requested for loading
    //      before. Used to prevent the same module being loaded twice
    // - l: string, Location, the url location of this module
    // - t: string, Text, the text content of the module
    // - e: booleany, Error, truthy if there was an error (probably a 404) loading the module
    // - n: module object, Next, instead of using this module, use the object
    //      pointed to by this property. Used for dependencies in other packages
    // - exports, object, the exports of the module!
    function getModule(location) {
        return MODULES[location] = MODULES[location] || {l: location};
    }

    function deepLoad(module, callback, parentLocation, path, dep) {
        if (module.g) {
            return callback(module.e, module);
        }

        var location = module.g = module.l;

        var request = new XMLHttpRequest();
        request.onload = function (text, deps, count) {
            if (request.status == 200) {
                text = module.t = request.response;
                deps = {};
                text.replace(/(?:^|[^\w\$_.])require\s*\(\s*["']([^"']*)["']\s*\)/g, function (_, id) {
                    deps[id] = true;
                });
                deps = Object.keys(deps);
                count = deps.length;
                function loaded() {
                    // We call loaded straight away below in case there
                    // are no dependencies. Putting this check first
                    // and the decrement after saves us an `if` for that
                    // special case
                    if (!count--) {
                        callback(undefined, module);
                    }
                }
                deps.map(function (dep) {
                    deepLoad(
                        getModule(resolve(module.l, dep)),
                        loaded,
                        // If it doesn't begin with a ".", then we're searching
                        // node_modules, so pass in the info to make this
                        // possible
                        dep[0] != "." ? location : undefined ,
                        "/../",
                        dep
                    );
                });
                loaded();
            } else {
                module.e = request;
                // parentLocation is only given if we're searching in node_modules
                if (parentLocation) {
                    // Recurse up the tree trying to find the dependency
                    // (generating 404s on the way)
                    deepLoad(module.n = getModule(resolve(parentLocation + (path += "../"), dep)), callback, parentLocation, path, dep);
                } else {
                    callback(request, module);
                }
            }
        };
        request.open("GET", location + ".js", true);
        request.send();
    }

    function getExports(module) {
        if (module.n) {
            return getExports(module.n);
        }

        if (!module[tmp]) {
            globalEval("(function(require,exports,module){"+module.t+"\n})")(
                function require (id) {
                    return getExports(MODULES[resolve(module.l, id)]);
                }, // require
                module[tmp] = {}, // exports
                module // module
            );
        }

        return module[tmp];
    }

    function R(id, callback) {
        deepLoad(getModule(resolve(location, id)), function (_, module) {
            id = getExports(module);
            if (callback) {
                callback(id);
            }
        });
    }

    tmp = document.querySelector("script[data-main]");
    if (tmp) {
        R(tmp.getAttribute("data-main"));
    }
    tmp = "exports";

    return R;

}(this, document));
